### PR TITLE
 Allow removing the client from the selected set in EpollDispatcher.unregister

### PR DESCRIPTION
**Pull request by nemanja-boric-sociomantic**
`EpollSelectDispatcher.unregister` now accepts an optional flag
(`remove_from_selected_set`) which will remove the client from the
selected set if it's in there. This will allow other ISelectClients to
safely destroy other ISelectClients from their handle method
(previously, this was unsafe because even unregistering the client
would just remove it from the all next epoll cycles, but not from the
other one, if the client's fd reported that it's ready to be handled).

The assumption taken in this PR is that the branch predictor will make most of the branches in the `SelectedKeysHandler.opCall` no-ops, or that they have very little cost compared with the IO that is done, so let's try this before taking alternative more complicated approaches.

### CODE REVIEW

File: src/ocean/io/select/selector/SelectedKeysHandler.d [hunk 1]
```diff
@@ -68,7 +68,7 @@ class SelectedKeysHandler: ISelectedKeysHandler
 
     ***************************************************************************/
 
-    public alias int delegate ( ISelectClient client ) UnregisterDg;
+    public alias int delegate ( ISelectClient client, bool = false ) UnregisterDg;
```
Comment (david-eckardt-sociomantic) [inline src/ocean/io/select/selector/SelectedKeysHandler.d]: This compiles?!? Anyway, a parameter name would be lovely.

File: src/ocean/io/select/EpollSelectDispatcher.d [hunk 2]
```diff
@@ -382,13 +387,27 @@ public class EpollSelectDispatcher : IEpollSelectDispatcherInfo
...
+                if (remove_from_selected_set)
+                {
+                    scope match_client_ptr = (Const!(epoll_event_t) entry) {
+                        return entry.data.ptr == cast(void*)client;
+                    };
+
+                    auto index = this.selected_set.findIf(match_client_ptr);
+                    if (index < this.selected_set.length)
+                    {
+                        this.selected_set[index] = epoll_event_t.init;
```
Comment (david-eckardt-sociomantic) [inline src/ocean/io/select/EpollSelectDispatcher.d]: Maybe you could remove the array element i.e.
```d
this.selected_set[index] = this.selected_set[$ - 1];
this.selected_set = this.selected_set[0 .. $ - 1];
```
?
Also, just `foreach (selected_client; this.selected_set)` is probably less code (a matter of taste, though).

Comment (nemanja-boric-sociomantic) [reply src/ocean/io/select/EpollSelectDispatcher.d]: Removing the array element is problematic because of two things:

1. When this is running, there's a good chance that `SelectedKeysHandler` is already doing `foreach (key; selected_keys)`, so we need to be very careful about changing the array while iterating.

2. How the things are setup at the moment, `EpollSelectDispatcher` has the array that both `EpollSelectDispatcher` and `SelectedKeysHandler` slice. Changing the `this.selected_set` slice will not reflect on the slices in `SelectedKeysHandler`, so we need to add the method to the `SelectedKeysHandler` to remove this array from it's slice, but also this will have to change the array stored here.

Since the only penalty here is checking with the null, and which would always pass, except for very obscure cases where the `ISelectClient` instances have the scoped lifetime, I don't think this is worth it.

Comment (david-eckardt-sociomantic) [reply src/ocean/io/select/EpollSelectDispatcher.d]: > Maybe you could remove the array element

As discussed verbally, this would interfere with the iteration in `SelectedKeysHandler.opCall` so never mind.

File: src/ocean/io/select/EpollSelectDispatcher.d [hunk 3]
```diff
@@ -791,9 +800,9 @@ public class EpollSelectDispatcher : IEpollSelectDispatcherInfo
 
                 version ( EpollCounters ) if ( n == 0 ) this.counters.timeouts++;
 
-                auto selected_set = events[0 .. n];
+                this.selected_set = events[0 .. n];
 
-                this.handle(selected_set, this.unhandled_exception_hook);
+                this.handle(this.selected_set, this.unhandled_exception_hook);
```
Comment (david-eckardt-sociomantic) [inline src/ocean/io/select/EpollSelectDispatcher.d]: Maybe set `this.selected_set = null` here?

Comment (nemanja-boric-sociomantic) [reply src/ocean/io/select/EpollSelectDispatcher.d]: Yes, thank you!

**Comment by nemanja-boric-sociomantic** [comment]
Updated with `scope (exit) this.selected_set = null`

File: src/ocean/io/select/EpollSelectDispatcher.d [hunk 4]
```diff
@@ -791,9 +800,9 @@ public class EpollSelectDispatcher : IEpollSelectDispatcherInfo
 
                 version ( EpollCounters ) if ( n == 0 ) this.counters.timeouts++;
 
-                auto selected_set = events[0 .. n];
+                this.selected_set = events[0 .. n];
```
Comment (gavin-norman-sociomantic) [inline src/ocean/io/select/EpollSelectDispatcher.d]: Should this actually be `this.events`, too?

File: src/ocean/io/select/EpollSelectDispatcher.d [hunk 5]
```diff
@@ -367,6 +369,9 @@ public class EpollSelectDispatcher : IEpollSelectDispatcherInfo
 
        Params:
             client = client to unregister
+            remove_from_selected_set = if true, removes the client from the
```
Comment (gavin-norman-sociomantic) [inline src/ocean/io/select/EpollSelectDispatcher.d:372]: Release notes: new feature.

File: src/ocean/io/select/EpollSelectDispatcher.d [hunk 6]
```diff
@@ -367,6 +369,9 @@ public class EpollSelectDispatcher : IEpollSelectDispatcherInfo
...
+                selected_set that we could be currently iterating. This will allow
```
Comment (gavin-norman-sociomantic) [inline src/ocean/io/select/EpollSelectDispatcher.d]: Small tweak / expansion:
"if true, removes the client from the selected set that may be currently being iterated over. This guarantees that the unregistered client's handle method will not be subsequently called by the selector. The client may thus be safely destroyed after unregistering."

File: src/ocean/io/select/EpollSelectDispatcher.d [hunk 7]
```diff
@@ -382,13 +387,27 @@ public class EpollSelectDispatcher : IEpollSelectDispatcherInfo
...
     {
         try if (client.is_registered)
         {
             scope (success)
             {
                 this.registered_clients -= client;
+
+                if (remove_from_selected_set)
+                {
+                    scope match_client_ptr = (Const!(epoll_event_t) entry) {
```
Comment (gavin-norman-sociomantic) [inline src/ocean/io/select/EpollSelectDispatcher.d]: What's the reason for this unusual syntax, as opposed to just making this a normal nested function?

File: src/ocean/io/select/EpollSelectDispatcher.d (hunk 2 above)
Comment (gavin-norman-sociomantic) [reply src/ocean/io/select/EpollSelectDispatcher.d]: Maybe add a comment that the item is not actually removed from the set, but set to null.

File: src/ocean/io/select/selector/SelectedKeysHandler.d [hunk 8]
```diff
@@ -122,7 +122,10 @@ class SelectedKeysHandler: ISelectedKeysHandler
     {
         foreach (key; selected_set)
         {
-            this.handleSelectedKey(key, unhandled_exception_hook);
+            if (key != epoll_event_t.init)
```
Comment (gavin-norman-sociomantic) [inline src/ocean/io/select/selector/SelectedKeysHandler.d]: Worth adding a comment explaining why an entry could be null.

File: integrationtest/unregisterclient/main.d [hunk 9]
```diff
@@ -0,0 +1,123 @@
...
+            auto initializer = client.classinfo.init;
+        }
+
+        ubyte* ptr = cast(ubyte*)client;
+        ptr[0..initializer.length] = 0;
+    }
+
+    private bool handler1()
+    {
+        count++;
```
Comment (gavin-norman-sociomantic) [inline integrationtest/unregisterclient/main.d]: `this.` :)

File: integrationtest/unregisterclient/main.d [hunk 10]
```diff
@@ -0,0 +1,123 @@
...
+        // unregister and delete the other instance
+        this.epoll.unregister(this.event2, true);
+        corruptEventObject(this.event2);
+
+        return false;
+    }
+
+    private bool handler2()
+    {
+        count++;
```
Comment (gavin-norman-sociomantic) [inline integrationtest/unregisterclient/main.d]: `this.` :)

File: integrationtest/unregisterclient/main.d [hunk 11]
```diff
@@ -0,0 +1,123 @@
...
+        ptr[0..initializer.length] = 0;
+    }
+
+    private bool handler1()
+    {
+        count++;
+
+        // unregister and delete the other instance
+        this.epoll.unregister(this.event2, true);
+        corruptEventObject(this.event2);
```
Comment (gavin-norman-sociomantic) [inline integrationtest/unregisterclient/main.d]: `this.` :)

File: integrationtest/unregisterclient/main.d [hunk 12]
```diff
@@ -0,0 +1,123 @@
...
+        return false;
+    }
+
+    private bool handler2()
+    {
+        count++;
+
+        // unregister and delete the other instance
+        this.epoll.unregister(this.event1, true);
+        corruptEventObject(this.event1);
```
Comment (gavin-norman-sociomantic) [inline integrationtest/unregisterclient/main.d]: `this.` :)

File: integrationtest/unregisterclient/main.d [hunk 13]
```diff
@@ -0,0 +1,123 @@
...
+
+import ocean.io.device.File;
+import ocean.io.select.EpollSelectDispatcher;
+import ocean.io.select.client.SelectEvent;
+import ocean.util.app.DaemonApp;
+import ocean.util.test.DirectorySandbox;
+import ocean.core.Test;
+
+/// Main application class
+class EpollTestApp : DaemonApp
```
Comment (gavin-norman-sociomantic) [inline integrationtest/unregisterclient/main.d]: Why is this a daemon app btw? You don't seem to be using any of its facilities.

**Comment by gavin-norman-sociomantic** [comment]
Even though it contains new features, this ought to be in a patch release, as it's required to fix a segfault bug in an application.

File: src/ocean/io/select/EpollSelectDispatcher.d (hunk 7 above)
Comment (david-eckardt-sociomantic) [reply src/ocean/io/select/EpollSelectDispatcher.d]: `scope`?

Comment (gavin-norman-sociomantic) [reply src/ocean/io/select/EpollSelectDispatcher.d]: Yeah. I've never seen anyone declare a nested function like that. I'm probably missing something.

Comment (david-eckardt-sociomantic) [reply src/ocean/io/select/EpollSelectDispatcher.d]: To be honest, me neither. I wonder when one should use `scope` explicitly an when not.

Comment (nemanja-boric-sociomantic) [reply src/ocean/io/select/EpollSelectDispatcher.d]: If you feel this is the unusual syntax, you're probably right, I'll change it to become the nested function.

**Comment by nemanja-boric-sociomantic** [comment]
> Even though it contains new features, this ought to be in a patch release, as it's required to fix a segfault bug in an application.

Good point.

File: integrationtest/unregisterclient/main.d (hunk 13 above)
Comment (nemanja-boric-sociomantic) [reply integrationtest/unregisterclient/main.d]: It's just the easiest way for me to write these tests - just copy the boilerplate from another test. I'll make it non-daemon app.

**Comment by mihails-strasuns-sociomantic** [comment]
This is similar to TimerSet/TimeoutManager problem we had earlier with task system, right?

**Comment by nemanja-boric-sociomantic** [comment]
Yes, it's pretty much the same problem.

File: integrationtest/unregisterclient/main.d (hunk 13 above)
Comment (gavin-norman-sociomantic) [reply integrationtest/unregisterclient/main.d]: I mean, do you need an app class at all?

Comment (nemanja-boric-sociomantic) [reply integrationtest/unregisterclient/main.d]: No, that's what I meant by making it non-daemon app - I'll remove all classes.

**Comment by nemanja-boric-sociomantic** [comment]
Updated with @gavin-norman-sociomantic 's comments + rebase to v3.6.x

File: src/ocean/io/select/EpollSelectDispatcher.d [hunk 14]
```diff
@@ -800,7 +830,9 @@ public class EpollSelectDispatcher : IEpollSelectDispatcherInfo
 
                 version ( EpollCounters ) if ( n == 0 ) this.counters.timeouts++;
 
-                this.selected_set = events[0 .. n];
+                this.selected_set = this.events[0 .. n];
+                scope (exit)
+                    this.selected_set = null;
```
Comment (gavin-norman-sociomantic) [inline src/ocean/io/select/EpollSelectDispatcher.d]: I think these three lines should be in the first commit.

Comment (nemanja-boric-sociomantic) [reply src/ocean/io/select/EpollSelectDispatcher.d]: Correct, I messed up the fixup.

Comment (nemanja-boric-sociomantic) [reply src/ocean/io/select/EpollSelectDispatcher.d]: Forgot to mention, this is moved to the right commit.

File: relnotes/unregister.feature.md [hunk 15]
```diff
@@ -0,0 +1,9 @@
+### Allow removing the client from the selected set in EpollDispatcher.unregister
```
Comment (gavin-norman-sociomantic) [inline relnotes/unregister.feature.md]: `EpollSelectDispatcher.unregister`

File: relnotes/unregister.feature.md [hunk 16]
```diff
@@ -0,0 +1,9 @@
...
+
```
Comment (gavin-norman-sociomantic) [inline relnotes/unregister.feature.md]: Should also list the names of affected modules. (`ocean.io.select.EpollSelectDispatcher`, in this case.)

File: relnotes/unregister.feature.md [hunk 17]
```diff
@@ -0,0 +1,9 @@
...
+`EpollSelectDispatcher.unregister` now accepts an optional flag
+(`remove_from_selected_set`) which will remove the client from the selected set
+if it's in there. This will allow other `ISelectClient`s to safely destroy
+other `ISelectClient`s from their handle method (previously, this was unsafe
```
Comment (gavin-norman-sociomantic) [inline relnotes/unregister.feature.md]: Maybe reword the section in (). It's quite unclear.

(You could probably copy some of the doc from the `unregister` method, say.)

File: src/ocean/io/select/EpollSelectDispatcher.d [hunk 18]
```diff
@@ -382,13 +389,36 @@ public class EpollSelectDispatcher : IEpollSelectDispatcherInfo
...
+                    {
+                        return entry.data.ptr == cast(void*)client;
+                    }
+
+                    auto index = this.selected_set.findIf(&entry_to_client);
+                    // Instead of removing the array entry, we'll just invalidate
+                    // it. This is to avoid problems with both the fact that
+                    // SelectedKeysHandler might be foreach-iterating over
+                    // this array at this time and the fact that shrinking
+                    // the slice owned by EpollSelectDispatcher wouldn't shink
```
Comment (gavin-norman-sociomantic) [inline src/ocean/io/select/EpollSelectDispatcher.d]: Typo shink.

**Comment by nemanja-boric-sociomantic** [comment]
Updated.

**Review by gavin-norman-sociomantic — APPROVED** [review]

**Comment by gavin-norman-sociomantic** [comment]
@david-eckardt-sociomantic please merge if you approve.
